torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1247,6 +1261,13 @@ void ProcessGroupNCCL::workCleanupLoop() {
       auto& work = *it;
       work.checkAndSetException();
       bool timedOut = work.checkTimeout();
+      c10::optional<std::thread> dumpingDebugInfo;
```

Comment:
A case we need to handle is when a process is the cause of a timeout: it hasn't started work, so others are blocked. In this case the workMetaList_ will be empty, and this code won't run.

I think we likely need a check where `heartbeat_++` is: 

```
if (last_time_we_removed_something_from_metaworkList - now > watchdog_timeout) {
   if (last_time_we_polled_for_dump_timeout > heartbeat_timeout) {
       if (timeout_dump is set in store) {
            <do a local dump> 
       } 
   }
}
```

Reply:
ok, you are talking about the case when the heartbeat is still increasing, but somehow the `workMetaList_` keeps being empty since last time removed?

I want to confirm one thing, do we perform the poll everytime we loop through each Work?

Reply:
The if statement is _outside_ of the loop over work. The idea that we should poll once very heartbeat_timeout seconds only if we haven't seen any work start or complete in the last watchdog_timeout seconds.  This is essentially just an optimization of always polling the store every heartbeat_timeout seconds: we know no one could have timed out if we have recently finished some nccl work, so there is no need to check the store.